### PR TITLE
[SelectionDAG] Add space-optimized forms of OPC_CheckComplexPat

### PR DESCRIPTION
**Pull request by wangpc-pp**
We record the usage of each `ComplexPat` and sort the `ComplexPat`s
by usage.

For the top 8 `ComplexPat`s, we will emit a `OPC_CheckComplexPatN`
to save one byte.

Overall this reduces the llc binary size with all in-tree targets by
about 89K.

### CODE REVIEW

**Comment by wangpc-pp** [comment]
Ping.

**Comment by wangpc-pp** [comment]
Ping.

**Review by ilovepi — APPROVED** [review]
This seems fine to me, and in line with the earlier patches. Would you mind adding some documentation about the change somewhere though? I feel like we should at least have a meta comment somewhere noting that the first 8 numbers are special.

You probably want to wait for one of the code owners to give the LGTM before landing, though.

**Comment by wangpc-pp** [comment]
> This seems fine to me, and in line with the earlier patches. Would you mind adding some documentation about the change somewhere though? I feel like we should at least have a meta comment somewhere noting that the first 8 numbers are special.

Thanks! I'm working on adding some comments to the opcodes and will create a PR later.

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp [hunk 1]
```diff
@@ -652,8 +655,13 @@ EmitMatcher(const Matcher *N, const unsigned Indent, unsigned CurrentIdx,
   case Matcher::CheckComplexPat: {
     const CheckComplexPatMatcher *CCPM = cast<CheckComplexPatMatcher>(N);
     const ComplexPattern &Pattern = CCPM->getPattern();
-    OS << "OPC_CheckComplexPat, /*CP*/" << getComplexPat(Pattern) << ", /*#*/"
-       << CCPM->getMatchNumber() << ',';
+    unsigned PatternNo = getComplexPat(Pattern);
+    if (PatternNo < 8)
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:659]: Braces

File: llvm/utils/TableGen/DAGISelMatcherGen.cpp [hunk 2]
```diff
@@ -56,7 +56,7 @@ static MVT::SimpleValueType getRegisterValueType(Record *R,
 namespace {
   class MatcherGen {
     const PatternToMatch &Pattern;
-    const CodeGenDAGPatterns &CGP;
+    CodeGenDAGPatterns &CGP;
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherGen.cpp:59]: Seems unfortunate to lose const for this

File: llvm/utils/TableGen/DAGISelMatcherGen.cpp [hunk 3]
```diff
@@ -602,6 +601,7 @@ bool MatcherGen::EmitMatcherCode(unsigned Variant) {
 
     // Emit a CheckComplexPat operation, which does the match (aborting if it
     // fails) and pushes the matched operands onto the recorded nodes list.
+    CGP.increaseComplexPatternUsage(CP);
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherGen.cpp:604]: Could you have done this before the actual emission to keep the const during emission?

Comment (wangpc-pp) [reply llvm/utils/TableGen/DAGISelMatcherGen.cpp]: Thanks!
I have changed the implementation, now we don't need to record the usage when creating `XXXMatcher` nodes. Instead, we will do statistic just before emission (with a single pass iteration over the whole MatcherTable). I do this for two reasons:
* The previous implementation is too intrusive, and, as you see, we lost the const qualifier.
* There are some optimizations after creating `XXXMatcher` nodes and some nodes may be deleted, so we should do the statistic after optimizations.

The implementation is much easier to review now.

**Review by arsenm — APPROVED** [review]

**Pull request merged**